File: python/trie/prefix-tree.py
class Node:
    def __init__(self):
        self.arr = [None] * 26
        self.isWord = False

class PrefixTree:

    def __init__(self):
        self.trie = Node()
    def getIdx(self, c: str) -> int:
        return ord(c) - 97
    def insert(self, word: str) -> None:
        curr = self.trie
        for c in word:
            idx = self.getIdx(c)
            if not curr.arr[idx]:
                curr.arr[idx] = Node() 
            curr = curr.arr[idx]
        curr.isWord = True
    def search(self, word: str) -> bool:
        curr = self.trie
        for c in word:
            idx = self.getIdx(c)
            if not curr.arr[idx]:
                return False
            curr = curr.arr[idx]
        return curr.isWord
    def startsWith(self, prefix: str) -> bool:
        curr = self.trie
        for c in prefix:
            idx = self.getIdx(c)
            if not curr.arr[idx]:
                return False
            curr = curr.arr[idx]
        return True
        

File: go/trie/prefix.tree.go
package main

type Node struct {
	children [26]*Node
	isWord   bool
}

type Trie struct {
	root Node
}

func Constructor() Trie {
	return Trie{root: Node{}}
}

func (t *Trie) GetIdx(c rune) int {
	return int(c) - int('a')
}

func (t *Trie) Insert(word string) {
	curr := &t.root

	for _, c := range word {
		idx := t.GetIdx(c)

		if curr.children[idx] == nil {
			curr.children[idx] = &Node{}
		}
		curr = curr.children[idx]
	}

	curr.isWord = true
}

func (t *Trie) Search(word string) bool {
	curr := &t.root

	for _, c := range word {
		idx := t.GetIdx(c)

		if curr.children[idx] == nil {
			return false
		}
		curr = curr.children[idx]
	}
	return curr.isWord
}

func (t *Trie) Startswith(word string) bool {
	curr := &t.root

	for _, c := range word {
		idx := t.GetIdx(c)

		if curr.children[idx] == nil {
			return false
		}
		curr = curr.children[idx]
	}
	return true
}
